ML-1913: Redis based nosql driver

assaf758 opened the pull request:
this replaces Alex PR at https://github.com/mlrun/storey/pull/352

Commit:
WIP: testing _save_key()

Commit:
Run hset on the event loop in a threadpool

Commit:
Fix lint issue

Commit:
Fix lint errors

Commit:
Use a default redis URL overridable by env

Commit:
Use fakeredis instead of travis redis

Commit:
Pin the correct version of fakeredis

Commit:
Fix version number

Commit:
Skip aggregation keys returned from HGETALL

Commit:
Remove unused variable

Commit:
WIP - README example with Redis

Commit:
Fix append to stream operation

Commit:
WIP - porting integration tests, driver logic

Commit:
WIP

Commit:
Progress on passing integration tests

Commit:
Make sure we initialize a new list, derp!

Commit:
Fix float/integer expectations

Commit:
Delete manual test

Commit:
Update tests to match new expected values from upstream

Commit:
Rename the redis aggregations integration test

Commit:
Always return aggregations as floats

Commit:
Fix return type for missing aggregations, additional data

Commit:
Fix lint error

Commit:
Add more comments

Commit:
Fix redis driver tests

Commit:
Remove unnecessary 'import' statements to make lint happy

Commit:
Make redis driver work with byte arrays

Commit:
LUA scripting - version 1

Commit:
Remove prints

Commit:
Add lupa (lua wrapper) to requirements.txt

Commit:
Lint fixes

Commit:
Add verifier for redis url

Commit:
Remove .gitignore and Pipfile

Commit:
Small fixes following gtopper comments

assaf758 committed:
remove ":" from the key-prefix

assaf758 committed:
add RedisNoSqlTarget

assaf758 committed:
adopt tests to new key

assaf758 committed:
address topper comments

assaf758 committed:
move RedisDriver to redis_driver.py

assaf758 committed:
REDIS_ENDPOINT => REDIS_API

assaf758 committed:
address gal commments

assaf758 committed:
webapi=>redis_url

assaf758 committed:
_convert_redis_value_to_python_obj: remove try/except clause

assaf758 committed:
remove RedisNoSqlTarget (mlrun to use NoSqlTarget instead)

assaf758 committed:
make _thread_pool a class var

assaf758 committed:
remove testing code

assaf758 committed:
fix test_redis_driver_read to use NoSqlTarget()

gtopper commented on integration/test_aggregation_with_redis.py:
```diff
@@ -0,0 +1,1853 @@
+import asyncio
```
Do we really have to duplicate `integration/test_aggregation_integration.py`? Can't we just parameterize the suite?

assaf758 replied:
Done

gtopper commented on setup.py:
```diff
@@ -27,6 +27,9 @@ def load_deps(file_name):
     "s3": ["s3fs~=0.5"],
     "az": ["adlfs~=0.5"],
     "kafka": ["kafka-python~=2.0"],
+    "redis": ["redis~=3.5"],
+    "redis-cluster": ["redis-py-cluster~=2.1.3"],
+    "lupa": ["lupa~=1.13"],
```
Any reason that redis and lupa are pinned to a major version and redis-cluster is pinned to a minor version?

assaf758 replied:
no reason, fixed

gtopper commented on tests/test_flow.py:
```diff
@@ -407,6 +408,9 @@ def test_write_parquet_read_parquet_partitioned(tmpdir):
 
 from unittest.mock import MagicMock
+from storey.redis_driver import RedisDriver
+from storey.aggregations import AggregateByKey
+from storey.dtypes import FieldAggregator, FixedWindows, SlidingWindows
```
These imports are not used.

assaf758 replied:
removed

gtopper commented on tests/test_flow.py:
```diff
@@ -3324,3 +3328,68 @@ def test_rename():
     assert termination_result == [{'b': 1, 'c': 3, 'e': 5}]
+
+
+def test_redis_driver_write(redis):
```
This file is for unit tests. These are actually integration tests when `REDIS_URL` is set.

assaf758 replied:
well, we do have V3io driver running on other tests in this file - what's the difference?
anyway, can move it to integration tests if you prefer.

gtopper replied:
There is no V3IO driver running in `test_flow.py`, because it's a unit test suite. Ideally fake redis will run in unit tests and real redis in integration tests, and without code duplication of course :).

assaf758 replied:
moved redis specific tests to out of test_flow.py and into new integration file test_redis_specific.py

gtopper commented on storey/utils.py:
```diff
@@ -304,3 +304,11 @@ def find_filters(partitions_time_attributes, start, end, filters, filter_column)
     _find_filter_helper(partitions_time_attributes, end, "<", "<=", first_uncommon, filters, filter_column)
+
+
+def drop_reserved_columns(df):
```
This function is added but never used.

assaf758 replied:
removed

gtopper commented on storey/table.py:
```diff
@@ -1075,6 +1090,9 @@ def __init__(self, max_value=None, set_data=None, set_time=None):
         self._set_value(self.value + value * value)
 
+    def aggregate_lua_script(self, vl1, vl2):
+        return f'{vl1}+{vl2}'
```
Hm, this doesn't look like the right implementation for sum of squares. Unless I'm missing something, we should add a test that catches this.

assaf758 replied:
the expected Sqr math function (x*x) is done in the aggregate function.
the aggregate_lua_script() is used as component for building the lua script. It is called with vl1 containing the old value and vl2 is the square of the value to be added.
so there is no issue here (and there are tests covering this functionality)

gtopper commented on storey/table.py:
```diff
@@ -1064,6 +1076,9 @@ def __init__(self, max_value=None, set_data=None, set_time=None):
         self._set_value(self.value + 1)
 
+    def aggregate_lua_script(self, vl1, vl2):
+        return f'{vl1}+{vl2}'
```
This doesn't look like the right implementation for count. Unless I'm missing something, we should add a test that catches this.

assaf758 replied:
similar explanation to SqrValue()

gtopper commented on storey/table.py:
```diff
@@ -1042,6 +1048,9 @@ def aggregate(self, time, value):
     def get_update_expression(self, old):
         return f'max({old}, {self.value})'
 
+    def aggregate_lua_script(self, vl1, vl2):
+        return f'type({vl1}) == "number" and math.max({vl1},{vl2}) or {vl2}'
```
Not sure I understand this... Is this supposed to be the value or a condition? Some of the implementations have what seems to be a value, e.g. `f'{vl1}+{vl2}'`, and some have what seems to be a condition. Also the signature here doesn't match the one in the super class.

assaf758 replied:
neither a value or condition - it's part of lua script, to be joined with other parts of the script (see lua_script in _build_feature_store_lua_update_script).
Changed the super-class signature and added a comment explaining this method.

assaf758 committed:
addressing gal comments

assaf758 committed:
refactor and increase test coverage of redis to all integration tests

1. removed test_aggregation_with_redis.py. instead, every test that use fixtures "setup_teardown_test" or "setup_kv_teardown_test" will be parameterized by driver (Redis, V3io).
2. moved redis spcefic tests from test_flow.py to integration/test_redis_specific.py
3. added support for fakeredis in all tests

gtopper commented on storey/table.py:
```diff
@@ -970,6 +970,14 @@ def __init__(self, max_value=None, set_data=None, set_time=None):
         raise NotImplementedError()
 
+    def aggregate_lua_script(self, vl1, vl2):
+        '''
```
```suggestion
        """
```

assaf758 replied:
fixed

gtopper commented on storey/table.py:
```diff
@@ -970,6 +970,14 @@ def __init__(self, max_value=None, set_data=None, set_time=None):
+        '''
+        The aggregate_lua_script() method is used for creating part of a lua script, to be sent
+        to the Redis DB (see _build_feature_store_lua_update_script).
+        The math function (for example Sqr: x*x, increment: x+1) is implemented in the aggregate method.'''
```
```suggestion
        The math function (for example Sqr: x*x, increment: x+1) is implemented in the aggregate method.
        """
```

assaf758 replied:
fixed

gtopper commented on integration/test_aggregation_integration.py:
```diff
@@ -1550,7 +1550,11 @@ def map_multiply(x):
-    table = Table(setup_teardown_test, V3ioDriver(), flush_interval_secs=None)
+
+    if setup_teardown_test["driver_name"] == "RedisDriver":
+        pytest.skip() # TODO
```
Let's not forget this one.

assaf758 replied:
addressing this is still in progress.

assaf758 replied:
fixed

gtopper commented on integration/test_flow_integration.py:
```diff
@@ -58,13 +58,35 @@ async def get_shard_data(self, path):
+    from .integration_test_utils import get_redis_client
+    from storey import RedisDriver
+    table_name = setup_teardown_test["table_name"]
+    redis_key = ''.join(('storey-test:', table_name, key, ':static'))
```
```suggestion
    redis_key = f'storey-test:{table_name}{key}:static'
```

assaf758 replied:
fixed

gtopper commented on integration/test_flow_integration.py:
```diff
@@ -58,13 +58,35 @@ async def get_shard_data(self, path):
+    return {RedisDriver.convert_to_str(key): RedisDriver.convert_redis_value_to_python_obj(val) for key, val in values.items()}
+
+def get_key_all_attrs_test_helper(setup_teardown_test: dict, key:str):
+    result = None
```
This line is redundant.

assaf758 replied:
fixed

gtopper commented on integration/test_flow_integration.py:
```diff
@@ -58,13 +58,35 @@ async def get_shard_data(self, path):
 
 
+def _get_redis_kv_all_attrs(setup_teardown_test: dict, key:str):
+    from .integration_test_utils import get_redis_client
+    from storey import RedisDriver
+    table_name = setup_teardown_test["table_name"]
+    redis_key = ''.join(('storey-test:', table_name, key, ':static'))
+    redis_fake_server = setup_teardown_test["redis_fake_server"] if "redis_fake_server" in setup_teardown_test else None
+    values = get_redis_client(redis_fake_server= redis_fake_server).hgetall(redis_key)    
+    return {RedisDriver.convert_to_str(key): RedisDriver.convert_redis_value_to_python_obj(val) for key, val in values.items()}
+
+def get_key_all_attrs_test_helper(setup_teardown_test: dict, key:str):
+    result = None
+    if setup_teardown_test["driver_name"] == "RedisDriver":
+        result = _get_redis_kv_all_attrs(setup_teardown_test, key)
+    else:
+        response = asyncio.run(get_kv_item(setup_teardown_test["table_name"], key))
+        assert response.status_code == 200
+        result = response.output.item
+    return result
+
 def test_join_with_v3io_table(setup_kv_teardown_test):
-    table_path = setup_kv_teardown_test
+    if setup_kv_teardown_test["driver_name"] == "RedisDriver":
+        pytest.skip(msg = 'test not relevant for Redis')
```
So maybe we can have separate fixtures?

assaf758 replied:
Since we are thinking about obsoleting/depreciating JoinWithV3IOTable, do we really want to a fixture that "escapes" the driver parametrization?

gtopper commented on integration/test_flow_integration.py:
```diff
@@ -58,13 +58,35 @@ async def get_shard_data(self, path):
         Map(lambda x: x + 1),
         Filter(lambda x: x < 8),
-        JoinWithV3IOTable(V3ioDriver(), lambda x: x, lambda x, y: y['age'], table_path),
+        JoinWithV3IOTable(get_driver(setup_kv_teardown_test), lambda x: x, lambda x, y: y['age'], table_path),
```
Seems like an unnecessary change.

assaf758 replied:
fixed

gtopper commented on integration/test_flow_integration.py:
```diff
@@ -320,7 +343,7 @@ def test_join_by_key(setup_kv_teardown_test):
 
 def test_join_by_key_specific_attributes(setup_kv_teardown_test):
-    table = Table(setup_kv_teardown_test, V3ioDriver())
+    table = Table(setup_kv_teardown_test["table_name"], get_driver(setup_kv_teardown_test))
```
I think it would be nicer to have `setup_kv_teardown_test` return a tuple?
As in
```suggestion
    table_name, _ = setup_kv_teardown_test
```

gtopper commented on integration/test_flow_integration.py:
```diff
@@ -320,7 +343,7 @@ def test_join_by_key(setup_kv_teardown_test):
 
 def test_join_by_key_specific_attributes(setup_kv_teardown_test):
-    table = Table(setup_kv_teardown_test, V3ioDriver())
+    table = Table(setup_kv_teardown_test["table_name"], get_driver(setup_kv_teardown_test))
```
We use single quotes in storey, except for `"""` and when single quotes need to appear in the actual string.
```suggestion
    table = Table(setup_kv_teardown_test['table_name'], get_driver(setup_kv_teardown_test))
```

assaf758 replied:
removed by introducing TestContext getters.

gtopper commented on integration/test_redis_specific.py:
```diff
@@ -0,0 +1,63 @@
+    import os
+    import fakeredis
+    import redis as r
+    REDIS_URL = os.environ.get('REDIS_URL')
+    if REDIS_URL:
+        yield r.Redis.from_url(REDIS_URL)
```
```suggestion
    redis_url = os.environ.get('REDIS_URL')
    if REDIS_URL:
        yield r.Redis.from_url(redis_url)
```

assaf758 replied:
fixed

gtopper commented on integration/integration_test_utils.py:
```diff
@@ -68,40 +73,101 @@ def __init__(self, **kwargs):
-@pytest.fixture()
-def setup_teardown_test():
+def get_redis_client(redis_fake_server=None):
+    REDIS_URL = os.environ.get('REDIS_URL')
+    if REDIS_URL:
+        return r.Redis.from_url(REDIS_URL)
```
```suggestion
    redis_url = os.environ.get('REDIS_URL')
    if redis_url:
        return r.Redis.from_url(redis_url)
```

assaf758 replied:
fixed

gtopper commented on integration/integration_test_utils.py:
```diff
@@ -68,40 +73,101 @@ def __init__(self, **kwargs):
+        redis_fake_server = setup_teardown_test["redis_fake_server"] if "redis_fake_server" in setup_teardown_test else None
+        return redis_driver(redis_fake_server = redis_fake_server, *args, **kwargs)
+    else:
+        assert 0
```
```suggestion
        driver_name = setup_teardown_test["driver_name"]
        ...
        raise ValueError(f'Unsupported driver name "{driver_name}"')
```

assaf758 replied:
fixed

gtopper commented on integration/integration_test_utils.py:
```diff
@@ -68,40 +73,101 @@ def __init__(self, **kwargs):
+    test_params["driver_name"] = driver_name
+    test_params["table_name"] = table_name
+    if driver_name == "RedisDriver":
+        REDIS_URL = os.environ.get('REDIS_URL')
+        if not REDIS_URL:
+            test_params["redis_fake_server"] = fakeredis.FakeServer()
```
I'm starting to think this code is repeated a lot... Can we extract it?
```suggestion
        redis_url = os.environ.get('REDIS_URL')
        if not redis_url:
            test_params["redis_fake_server"] = fakeredis.FakeServer()
```

assaf758 replied:
done

gtopper commented on integration/integration_test_utils.py:
```diff
@@ -68,40 +73,101 @@ def __init__(self, **kwargs):
+    test_params["driver_name"] = driver_name
+    test_params["table_name"] = table_name
+    if driver_name == "RedisDriver":
+        REDIS_URL = os.environ.get('REDIS_URL')
+        if not REDIS_URL:
+            test_params["redis_fake_server"] = fakeredis.FakeServer()
```
Here as well.

assaf758 replied:
done

gtopper commented on integration/integration_test_utils.py:
```diff
@@ -68,40 +73,101 @@ def __init__(self, **kwargs):
+
+    if driver_name == "V3ioDriver":
+        asyncio.run(create_temp_kv(table_name))
+    elif driver_name == "RedisDriver":
+        create_temp_redis_kv(test_params)
+    else:
+        assert 0
 
     # Test runs
-    yield table_path
+    yield test_params
 
     # Teardown
-    asyncio.run(recursive_delete(table_path, V3ioHeaders()))
+    if driver_name == "V3ioDriver":
+        asyncio.run(recursive_delete(table_name, V3ioHeaders()))
+    elif driver_name == "RedisDriver":
+        remove_redis_table(table_name)
+    else:
+        assert 0
```
Here too we need to raise an error with what happened as mentioned above. Not just `assert False` (let alone 0...).

assaf758 replied:
fixed

gtopper commented on integration/integration_test_utils.py:
```diff
@@ -68,40 +73,101 @@ def __init__(self, **kwargs):
+    
+    test_params = {}
+    driver_name = request.param
+    test_params["driver_name"] = driver_name
+    test_params["table_name"] = table_name
```
I'm starting to think this way makes for too much clutter in test code... If we really need this as opposed to just a tuple result, can we create a class for it? It makes for more readable (and writable) client code.

assaf758 replied:
class created

gtopper commented on storey/drivers.py:
```diff
@@ -156,9 +156,11 @@ async def _save_key(self, container, table_path, key, aggr_item, partitioned_by_
         response = await self._v3io_client.kv.update(container, table_path, key, expression=update_expression,
                                                      condition=condition_expression,
                                                      raise_for_status=v3io.aio.dataplane.RaiseForStatus.never)
+
```
Let's not change the formatting in this file.

assaf758 replied:
fixed

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+class NeedsRedisAccess:
+    """Checks that params for access to Redis exist and are legal
+
+    :param webapi: URL to the web API (https or http). If not set, the REDIS_URL environment variable will be used.
```
I don't think this comment is correct for redis. Redis URLs have the "redis" scheme rather than "http" or "https", and are not referred to as "web API". So this seems like accidental c&p from v3io.

assaf758 replied:
fixed

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+    """Checks that params for access to Redis exist and are legal
+
+    :param webapi: URL to the web API (https or http). If not set, the REDIS_URL environment variable will be used.
+
```
```suggestion
```

assaf758 replied:
fixed

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+        redis_url = redis_url or os.getenv('REDIS_URL')
+        if not redis_url:
+            self._redis_url = None
+            print('Missing redis_url parameter or REDIS_URL environment variable. Using fakeredit instead')
```
Hm, we're using fakeredis in the implementation code? Isn't it only for tests?

assaf758 replied:
it is, fixed logic.

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+        async def inner_fn(*args, **kwargs):
+            loop = asyncio.get_event_loop()
+            partial_hset = partial(fn, *args, **kwargs)
+            return await loop.run_in_executor(RedisDriver._thread_pool, partial_hset)
```
Elsewhere in storey we pass `None` for the executor (thread pool is the implementation type) to get the default executor. Why do we need a dedicated parameterless executor here?

assaf758 replied:
since the redis driver 3.5 does not support asyncio, using multiple threads provided a mitigation.
I now moved to redis 4.3 which does support asyncio. I will evaluate moving to it later on (not in this PR).

gtopper replied:
This doesn't answer my question though... Why not use the default executor as is done elsewhere in storey?

assaf758 replied:
we can do that - but the code will run slower.
the reason for having multiple threads is to hide the latency of the call to Redis: the other thread will be blocked waiting for the redis call to complete while the main thread can handle other coroutines.

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+        self._aggregation_prefixes = (self._aggregation_attribute_prefix,
+                                      self._aggregation_time_attribute_prefix)
+
+    def asyncify(self, fn):
```
Method can be moved out of class or be made static.

assaf758 replied:
made static

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+        """Run a synchronous function asynchronously."""
+        async def inner_fn(*args, **kwargs):
+            loop = asyncio.get_event_loop()
+            partial_hset = partial(fn, *args, **kwargs)
```
Seems to me that `partial_hset` is a leftover from when this function was specific to the HSET operation.
```suggestion
            partial_func = partial(fn, *args, **kwargs)
```

assaf758 replied:
changed to partial_fn

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+
+    @staticmethod
+    def make_key(key_prefix, *parts):
+        return f"{key_prefix}{''.join([str(p) for p in parts])}"
```
```suggestion
        key = ''.join([str(p) for p in parts])
        return f'{key_prefix}{key}'
```
but hm, is it even correct to join the keys without a separator?

assaf758 replied:
it is - the table_path part contains '/' both at it's start and end.

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+            return f'{cls.TIMEDELTA_FIELD_PREFIX}{value.total_seconds()}'
+        else:
+            if isinstance(value, float):
+                if value == math.inf or value == -math.inf:
+                    # Use the shorter infinity form (inf, -inf) to save space.
+                    value = str(value)
```
```
>>> json.dumps(math.inf)
'Infinity'
>>> json.dumps(str(math.inf))
'"inf"'
```
The comment implies that `value = str(value)` is an optimization, but the code doesn't work without it. So need to remove or improve the comment.

What happens if the value is NaN?

assaf758 replied:
it's an optimization. If you want to remove it you have to do the on the function that translate back to python.
Aded NaN handling - good catch.

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+            return f'"{json.dumps(value)}"'
+        else:
+            if isinstance(value, float):
+                if value == math.inf:
+                    value = "math.huge"
+                elif value == -math.inf:
+                    value = "-math.huge"
```
Here too, are we sure NaN works?

assaf758 replied:
will be addressed in next commit

assaf758 replied:
fixed

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+            ret = timedelta(seconds=float(value.split(":")[1]))
+        elif value.startswith(cls.DATETIME_FIELD_PREFIX):
+            ret = datetime.fromtimestamp(float(value.split(":")[1]), tz=timezone.utc)
+        elif value == "inf":
+            ret = math.inf
+        elif value == "-inf":
+            ret = -math.inf
```
Here too, should we be concerned about NaN?

assaf758 replied:
will be addressed in next commit

assaf758 replied:
fixed

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+            ret = False
+        else:
+            # if value is a number, convert type to int / float
+            if value[0] in ('-', '+'):
+                value = value[1:]
+            if value.replace('.', '', 1).isdigit():
+                ret = int(value) if int(float(value)) == float(value) else float(value)
```
Am I missing something, or we just ignore the sign here?

assaf758 replied:
good catch - fixed

gtopper commented on tests/test_flow.py:
```diff
@@ -23,6 +23,7 @@
     ParquetTarget, QueryByKey, \
     TSDBTarget, Extend, SendToHttp, HttpRequest, NoSqlTarget, NoopDriver, Driver, Recover, V3ioDriver, ParquetSource
 from storey.flow import _ConcurrentJobExecution, Context, ReifyMetadata, Rename
+from integration.integration_test_utils import append_return
```
I'm pretty sure there's no need to change this file.

assaf758 replied:
fixed

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+            redis_key_prefix, aggr_item, partitioned_by_key, additional_data)
+        if not update_expression:
+            return
+        current_time = int(time.time_ns() / 1000)
```
Microseconds on purpose?

assaf758 replied:
occurs twice in the code so I assume yes :-)
what is the concern? suggest keeping the mtime at nano resolution?

gtopper replied:
Not really concerned about it, just wanted to make sure it's a deliberate choice :)

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+            update_expression = f'{update_expression}redis.call("HSET","{redis_key_prefix}","{self._mtime_name}",{current_time});return 1;'
+
+        redis_keys_involved.append(redis_key_prefix)
+        update_expression = f"""{update_expression}"""
```
This line seems redundant.

assaf758 replied:
removed

gtopper commented on storey/redis_driver.py:
```diff
@@ -0,0 +1,482 @@
+        redis_keys_involved.append(redis_key_prefix)
+        update_expression = f"""{update_expression}"""
+        update_ok = await self.asyncify(self.redis.eval)(update_expression, len(redis_keys_involved), *redis_keys_involved)
+        # should_raise_error = False
```
Dead code ☠️

assaf758 replied:
removed

assaf758 committed:
address gal's comments

assaf758 committed:
addressing gal comment: introduce TestContext class

assaf758 committed:
move to redis-py 4.3 which supports redis cluster and remove redis-py-cluster dependency

assaf758 committed:
handle NaN, test_write_read_first_last pass

assaf758 committed:
formatting

assaf758 committed:
allow for url rediss://

assaf758 committed:
fix redis property, add close method

gtopper commented on storey/drivers.py:
```diff
@@ -244,7 +246,6 @@ def _build_feature_store_update_expression(self, aggregation_element, additional
                     expressions.append(f'{name}={self._convert_python_obj_to_expression_value(value)}')
                 else:
                     expressions.append(f'REMOVE {name}')
-
```
No need to change formatting unnecessarily.

assaf758 replied:
fixed

gtopper commented on storey/drivers.py:
```diff
@@ -168,10 +169,12 @@ async def _save_key(self, container, table_path, key, aggr_item, partitioned_by_
             response = await self._v3io_client.kv.update(container, table_path, key, expression=update_expression,
                                                          condition=condition_expression,
                                                          raise_for_status=v3io.aio.dataplane.RaiseForStatus.never)
+
```
```suggestion
```

assaf758 replied:
fixed

gtopper commented on storey/drivers.py:
```diff
@@ -168,10 +169,12 @@ async def _save_key(self, container, table_path, key, aggr_item, partitioned_by_
                 await self._fetch_state_by_key(aggr_item, container, table_path, key)
             else:
                 should_raise_error = True
+
```
```suggestion
```

assaf758 replied:
fixed

gtopper commented on storey/drivers.py:
```diff
@@ -159,6 +159,7 @@ async def _save_key(self, container, table_path, key, aggr_item, partitioned_by_
         if response.status_code == 200:
             if aggr_item:
                 aggr_item.storage_specific_cache[self._mtime_header_name] = response.headers[self._mtime_header_name]
+
```
```suggestion
```

assaf758 replied:
fixed

assaf758 committed:
address gal's comments

assaf758 committed:
init self._redis to None if no redis_client provided at __init__

assaf758 committed:
export RedisType

assaf758 committed:
REDIS_URL => MLRUN_REDIS_URL

assaf758 committed:
rebase and add support for test_aggregate_multiple_keys_and_aggregationless_query

assaf758 committed:
remove Redis from "__init__.py" and create threadpool only on RedisDriver instansetiation

assaf758 committed:
use default executor
